browser/speedreader/speedreader_tab_helper.cc

```diff
@@ -33,25 +33,26 @@ base::WeakPtr<SpeedreaderTabHelper> SpeedreaderTabHelper::GetWeakPtr() {
 }
 
 SpeedreaderTabHelper::SpeedreaderTabHelper(content::WebContents* web_contents)
-    : content::WebContentsObserver(web_contents) {}
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<SpeedreaderTabHelper>(*web_contents) {}
 
 bool SpeedreaderTabHelper::IsSpeedreaderEnabled() const {
-  Profile* profile =
-      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  Profile* profile = Profile::FromBrowserContext(
+      const_cast<content::WebContents&>(GetWebContents()).GetBrowserContext());
```

Comment:
not sure I follow why would we need to get rid of `web_contents()`. All tab helpers in chromium main continue to use it, don't they?

Reply:
The upstream CL https://crrev.com/c/3327865 removed the default constructor for `WebContentsUserData` which forced us to change, in https://github.com/brave/brave-core/pull/11829/commits/b4908c6fb16d22a2e8d9d923581d58e58c18ec7a , how we construct objects for classes inheriting from it (e.g. like this one) and now there are cases where we can't keep using `web_contents()` anymore at construction time, simply because the object won't be correctly initialized if such call es made from a method that is -directly or indirectly- being called from the class constructor.

There are some obvious cases like `BraveWalletTabHelper`, where the `web_contents()` method would return a pointer that was manually managed by the class via a `raw_ptr` (see the [change in brave_wallet_tab_helper.h](https://github.com/brave/brave-core/pull/11829/commits/b4908c6fb16d22a2e8d9d923581d58e58c18ec7a#diff-47e29d98c9a932610c1550a2ad6bf34415b8740f811b14aefcb7525d6dfe700aL65)) and then some other ones that are less obvious where the `web_contents()` method being called is actually one coming from another superclass... which will only return valid data after the construction process has been completed.

As far as I could see looking into several of the many migrations referenced from crbug.com/1268914, the best way forward now when needing to access the `WebContents` from subclasses of `WebContentsUserData` is to call `WebContentsUserData::GetWebContents()`, which will always return a valid object even if called during construction time, so that's why the change in https://github.com/brave/brave-core/pull/11829/commits/b4908c6fb16d22a2e8d9d923581d58e58c18ec7a made the `web_contents() => GetWebContents()` switch in every single place like this one.

Now, it's possible that in this particular case calling `web_contents()` would be ok (i.e. as that would call `WebContentsObserver::web_contents()`), but I thought it would be good to be consistent and use the same method everywhere for every subclass of `WebContentsUserData`.

WDYT? Does this make sense? If not I can audit the changes and only do that for the cases it's strictly needed, although I fear that would be a bit confusing... :innocent:

Reply:
I want us to follow upstream conventions, I just indicated that I don't see mass migration to `GetWebContents()` in regular observers in upstream. Maybe you could link an example?

Reply:
@iefremov I'm so sorry for the confusion but **my reasoning from yesterday is not correct*... I was well on my way trying to provide you with a list of links to CLs to show what I was talking about based on the CLs landed for https://bugs.chromium.org/p/chromium/issues/detail?id=1268914, and I came up with "this list":
```
https://chromium-review.googlesource.com/c/chromium/src/+/3272814
https://chromium-review.googlesource.com/c/chromium/src/+/3276238
https://chromium-review.googlesource.com/c/chromium/src/+/3275544
https://chromium-review.googlesource.com/c/chromium/src/+/3278400
https://chromium-review.googlesource.com/c/chromium/src/+/3278376
https://chromium-review.googlesource.com/c/chromium/src/+/3286864
https://chromium-review.googlesource.com/c/chromium/src/+/3288450
https://chromium-review.googlesource.com/c/chromium/src/+/3328304
https://chromium-review.googlesource.com/c/chromium/src/+/3328565
```
... then it hit me: actually the only cases where upstream migrated to `GetWebContents()` is in cases like the one I mentioned yesterday for [BraveWalletTabHelper](https://github.com/brave/brave-core/pull/11829/commits/b4908c6fb16d22a2e8d9d923581d58e58c18ec7a#diff-47e29d98c9a932610c1550a2ad6bf34415b8740f811b14aefcb7525d6dfe700aL65), where we have no other option since we're removing a locally managed pointer to WebContents (usually a private `web_contents_` member).

Then I realized that the problems I had with the lifetime of WebContents objects during construction were not really related to using `WebContentsObserver::web_contents()` but because on an early version of my patch I actually forgot by mistake to keep calling the constructor of the `WebContentsObserver`  class from the subclasses' constructor... and that's where those nulls came from. Problem is that when I fixed that problem I didn't link it at the time with the `web_contents() -> GetWebContents()` massive replacement, and I got it all wrong.

Again, apologies for the confusion and thanks for pushing me into looking again into this, fixed in https://github.com/brave/brave-core/commit/bdfc2701f51dae3e7416850dddd12a042afd238c (which restores all of those unnecessary changes to the original code).

Reply:
gotcha, thanks for fixing it!